plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp

```diff
@@ -16,691 +19,327 @@ extern const char* const state_history_plugin_abi;
+            if(block_to_send) {
+               const fc::unsigned_int get_blocks_result_variant_index = block_to_send->is_v1_request ?
+                                                                        state_result(get_blocks_result_v1()).index() :
+                                                                        state_result(get_blocks_result_v0()).index();
+               co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(get_blocks_result_variant_index)));
+               co_await stream.async_write_some(false, boost::asio::buffer(fc::raw::pack(block_to_send->blocks_result_base)));
+
+               //accessing the _logs here violates the rule that those should only be accessed on the main thread. However, we're
+               // only calling get_unpacked_entry() on it which assumes the mutex is held by the locked_decompress_stream. So this is
+               // "safe" in some aspects but can deadlock
+               co_await write_log_entry(block_to_send->trace_stream, trace_log, block_to_send->this_block_num);
+               co_await write_log_entry(block_to_send->state_stream, chain_state_log, block_to_send->this_block_num);
+               if(block_to_send->is_v1_request)
+                  co_await write_log_entry(block_to_send->finality_stream, finality_data_log, block_to_send->this_block_num);
+
+               co_await stream.async_write_some(true, boost::asio::const_buffer());
+            }
+         }
+      });
    }
+
+private:
+   ///these items must only ever be touched by the session's strand
+   Executor                          strand;
+   coro_throwing_stream              stream;
+   coro_nonthrowing_steadytimer      wake_timer;
+   unsigned                          coros_running = 0;
+
+   ///these items must only ever be touched on the main thread
+   std::deque<bool>                  queued_status_requests;  //false for v0, true for v1
+
+   get_blocks_request_v0             current_blocks_request;
+   std::optional<bool>               current_blocks_request_v1_finality; //unset: current request is v0; set means v1; true/false is if finality requested
+   //current_blocks_request is modified with the current state; bind some more descriptive names to items frequently used
+   uint32_t&                         send_credits = current_blocks_request.max_messages_in_flight;
+   chain::block_num_type&            next_block_cursor = current_blocks_request.start_block_num;
+
+   chain::controller&                controller;
+   std::optional<state_history_log>& trace_log;
+   std::optional<state_history_log>& chain_state_log;
+   std::optional<state_history_log>& finality_data_log;
+
+   GetBlockID                        get_block_id;
+   GetBlock                          get_block;
+
+   ///these items might be used on either the strand or main thread
+   std::atomic_flag                  has_logged_exception;  //as is right now, this doesn't get used on anything but the strand
```

Comment:
I know it's really nitpicky but I really like one aspect of `atomic_flag`'s interface in that I can densely write
```c++
if(!has_logged_exception.test_and_set())
   ilog("printed once");
```
since increment operators are not defined on `bool` any longer, I _can't_ do something like
```c++
if(!has_logged_exception++)
   ilog("printed once");
```
instead it seems like I'm forced in to something like
```c++
if(!has_logged_exception) {
   has_logged_exception = true;
   ilog("printed once");
}
```
unless I want to get a little gnarly with something like
```c++
if(!has_logged_exception++)
   ilog("printed once"), has_logged_exception = true;
```
I almost want a "non atomic atomic_flag" I guess. In this case, `has_logged_exception` could also probably be a `uint64_t` and realistically never roll over either, allowing for usage of the increment operator and remaining dense

Reply:
In 8df3664 I moved a couple items that are only used in the session strand up in to the block commented by
https://github.com/AntelopeIO/spring/blob/8df36644a0eb3b02370f95e21f4bc0effe97c95f/plugins/state_history_plugin/include/eosio/state_history_plugin/session.hpp#L320
but I rolled back one of them in e984468 because I didn't catch that it caused a reorder warning.

It's probably for the better anyways, since I'd like for any user of the `session` to not expect the `on_done` to be called in any particular thread context.

I didn't change the `atomic_flag` to a bool because of the `test_and_set()` convenience I like :woman_shrugging: